internal/scraper/scraper_test.go

```diff
@@ -0,0 +1,457 @@
+	dto "github.com/prometheus/client_model/go"
+	promconfig "github.com/prometheus/common/config"
+	"github.com/prometheus/common/expfmt"
+	"github.com/stretchr/testify/require"
+)
+
+var updateGolden = flag.Bool("update-golden", false, "update golden files")
+
+func TestProber(t *testing.T) {
+	testcases := map[string]struct {
+		setup  func(t *testing.T) (string, func())
+		config bbeconfig.Module
+		prober prober.ProbeFn
+	}{
+		"ping": {
+			prober: prober.ProbeICMP,
+			setup: func(*testing.T) (string, func()) {
+				return "127.0.0.1", func() {}
+			},
+			config: bbeconfig.Module{
+				Prober:  "icmp",
+				Timeout: 2 * time.Second,
+				ICMP: bbeconfig.ICMPProbe{
+					IPProtocol:         "ip4",
+					IPProtocolFallback: false,
+				},
+			},
+		},
+
+		"http": {
+			prober: prober.ProbeHTTP,
+			setup: func(*testing.T) (string, func()) {
+				httpSrv := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+					w.WriteHeader(http.StatusOK)
+				}))
+				// httpSrv.EnableHTTP2 = true
+				httpSrv.Start()
+				return httpSrv.URL, httpSrv.Close
+			},
+			config: bbeconfig.Module{
+				Prober:  "http",
+				Timeout: 2 * time.Second,
+				HTTP: bbeconfig.HTTPProbe{
+					IPProtocol:         "ip4",
+					IPProtocolFallback: false,
+					HTTPClientConfig:   promconfig.HTTPClientConfig{},
+				},
+			},
+		},
+
+		"http_ssl": {
+			prober: prober.ProbeHTTP,
+			setup: func(*testing.T) (string, func()) {
+				httpSrv := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+					w.WriteHeader(http.StatusOK)
+				}))
+				// httpSrv.EnableHTTP2 = true
+				httpSrv.StartTLS()
+				return httpSrv.URL, httpSrv.Close
+			},
+			config: bbeconfig.Module{
+				Prober:  "http",
+				Timeout: 2 * time.Second,
+				HTTP: bbeconfig.HTTPProbe{
+					IPProtocol:         "ip4",
+					IPProtocolFallback: false,
+					HTTPClientConfig: promconfig.HTTPClientConfig{
+						TLSConfig: promconfig.TLSConfig{
+							InsecureSkipVerify: true,
+						},
+					},
+				},
+			},
+		},
+
+		"dns": {
+			prober: prober.ProbeDNS,
+			setup: func(*testing.T) (string, func()) {
+				dnsSrv, dnsAddr := startDNSServer(":0", "udp6", recursiveDNSHandler)
+				go func() {
+					err := dnsSrv.ActivateAndServe()
+					if err != nil {
+						// this should never happen, but log it if it does
+						t.Log(err)
+					}
+				}()
+				return dnsAddr.String(), func() {
+					err := dnsSrv.Shutdown()
+					if err != nil {
+						// this should never happen, but log it if it does
+						t.Log(err)
+					}
+				}
+			},
+			config: bbeconfig.Module{
+				Prober:  "dns",
+				Timeout: 2 * time.Second,
+				DNS: bbeconfig.DNSProbe{
+					IPProtocol:         "ip4",
+					IPProtocolFallback: true,
+					QueryName:          "example.com",
+					TransportProtocol:  "udp",
+				},
+			},
+		},
+
+		"tcp": {
+			prober: prober.ProbeTCP,
+			setup: func(t *testing.T) (string, func()) {
```

Comment:
nit: same here, let's move all this logic into a function.

Reply:
done